// encryptedcookies-paypal/lib/cookieIds.js
'use strict';

module.exports = {

    /** The cleartext name of the webscr login cookie */
    WEBSCR_LOGIN_COOKIE_NAME: 'user_session',

    /**
     * DEPRECATED!
     * The standard name of the Java session cookie.
     * We can't use the Java standard JSESSIONID, because webscr manages that
     * cookie due to its integration with VPS.
     */
    JSESSIONID: 'SPARTAJSESSIONID',

    /**
     * session cookie Next Generation.
     * have to keep both around for a while to be backwards compatible.
     */
    JSESSIONIDV2: 'SPARTAJSESSIONIDV2',

    /**
     * The cookie used to track user pref of language.
     * Format: plain text:  locale;country
     * Example: en_US;US
     */
    LANG: 'LANG',

    /**
     * The name of the cookie that indicates the control group that the user request belong to.
     */
    BETA_COOKIE_NAME: 'SEGM',

    /**
     * The cookie used in development mode to track service endpoint overrides
     */
    ENDPOINT_OVERRIDE: 'ENDPOINT_OVERRIDE',

    /**
     * The cookie used to confirm if cookies are disabled by the user.
     */
    COOKIE_CHECK: 's_cookie_check',

    /** Navigation cookie for last navigation state (LNS). */
    NAV_LNS: 'navlns',

    /** Navigation cookie for user percent. */
    NAV_USER_PERCENT: 'upct',

    /** Navigation cookie for virtual terminal enablement. */
    NAV_IS_VT_ENABLED: 'cookie_nav_is_vt_enabled',

    /** Navigation cookie for UUM enablement. */
    NAV_IS_UUM_ENABLED: 'cookie_nav_is_uum_enabled',

    LOGIN_EMAIL_COOKIE: 'login_email',

    VISITOR_ID_COOKIE: 'visitor_id',

    GUID_COOKIE: 'guid'

};
