// nemo-jaws/plugin/user/user-client.js
'use strict';
var Client = require('node-rest-client').Client,
    client = new Client(),
    baseUrl = "",
    lib = require("../../lib"),
    findByEmail = function(emailAddress, callback) {
        var getUserArgs = {
            "parameters": {
                "emailAddress": emailAddress
            },
            "headers": this.sharedHeader
        };
        client.methods.findByEmail(getUserArgs, lib.responseHandler(callback, "user.findByEmail"));
    },
    post = function(userArgs, callback) {
        var newUserArgs = {};

        newUserArgs.data = userArgs;
        newUserArgs.headers = this.sharedHeader;
        client.methods.post(newUserArgs, lib.responseHandler(callback, "user.post"));
    },
    put = function(userArgs, callback) {
        var putUserArgs = {
            "path": {
                "accountNumber": userArgs.accountNumber
            }
        };

        putUserArgs.data = userArgs;
        putUserArgs.headers = this.sharedHeader;
        client.methods.put(putUserArgs, lib.responseHandler(callback, "user.put"));
    },
    get = function(userArg, callback) {
        var getUserArgs = {
            "path": {
                "accountNumber": userArg.accountNumber
            },
            "headers": this.sharedHeader
        };
        client.methods.get(getUserArgs, lib.responseHandler(callback, "user.get"));

    };




module.exports = function() {
    this.configure = function(config) {
        this.stage = config.stage;
        this.sharedHeader = {
            "hostName": this.stage + ".qa.paypal.com",
            "Content-type": "application/json"
        };
		baseUrl = config.baseUrl;
		client.registerMethod("post", baseUrl + "/user", "POST");
		client.registerMethod("put", baseUrl + "/user/${accountNumber}", "PUT");
		client.registerMethod("get", baseUrl + "/user/${accountNumber}", "GET");
		client.registerMethod("findByEmail", baseUrl + "/user/findByEmail", "GET");
        //POST user create
        this.post = post.bind(this);
        //GET user by email
        this.findByEmail = findByEmail.bind(this);
        //GET user by account number
        this.get = get.bind(this);
        //PUT user update
        this.put = put.bind(this);
    };

};
