// mayfly/lib/constants.js
'use strict';

var MessageSenderType = {
    Dummy: 0,
    Client: 1,
    DirectoryServer: 2,
    StorageServer: 3,
    Persistent: 4
};

/**
 *  Mayfly constants
 */
var CONSTANTS = {
    magic: 0xDEEFCAFE,        // Magic code for Mayfly
    headerSize: 40,           // 40 bytes
    defaultLifeTime: 3600,    // 1 hr in seconds
    maxLifeTime: 259200,      // 3 days in seconds
    noLifeTimeUpdate: 0,      // Use original record lifetime
    noRetries: 3,             // No of retries
    defaultTimeout: 1200,      // 1200ms default SLA
    defaultConnectTimeout: 500,// 500ms default connect timeout
    defaultPort: 10368,
    maxKeyLen: 128,
    maxPayloadSize: 204800,   // 200kB
    maxSerialNumber: 65535,   // 2^16 - 2 byte 0xffff limit
    maxTimeStamp: 2147483648, // 2^31  - to enforce 4 bytes limit
    stageCryptKey: '53JweuLU+BimDBPnrFqvNYscm/I=',
    stageMacKey: '0Qcilz3zZxz0U+TomnvGZ1AuXpg='
};

var Direction = {
    unknown: 0,
    request: 1,
    response: 2
};

var DataTags = {
    applicationNameTag: 0xAB,
    operationMessageTag: 0xDD,
    namespaceTag: 0x01,
    keyDataTag: 0x02,
    recordInfoTag: 0x01,
    requestIdTag: 0x02,
    optionalDataTag: 0x04,
    payloadDataTag: 0x03,
    endTag: 0x00
};

/**
 * Operation Type tells the action - Create, Get, Update, Destroy...
 */
var OperationType = {
    NOP: 0,
    Create: 1,
    Get: 2,
    Update: 3,
    Destroy: 4,
    GetStats: 5,
    CommitWrite: 6,
    WriteData: 7,
    ReadData: 8,
    AbortWrite: 9,
    DeleteData: 10,
    Append: 11,
    Set: 12
};

/**
 *  Operation Status of the request
 */
var OperationStatus = {
    NO_ERROR: 0,
    BAD_MSG: 1,
    SERVICE_DENIED: 2,
    NO_KEY: 3,                 // storage serv only
    DUP_KEY: 4,
    DATA_EXPIRED: 5,
    OUT_OF_MEM: 6,
    BAD_PARAM: 7,
    RECORD_LOCKED: 8,          // storage serv only
    VERSION_TOO_OLD: 9,        // storage serv only
    NO_UNCOMMITTED: 10,        // storage serv only
    BAD_REQUEST_ID: 11,        // storage serv only
    NO_STORAGE_SERVER: 12,     // directory serv only
    DUPLICATE_REQUEST: 13,     // directory serv only
    STORAGE_SERV_TIMEOUT: 14,  // deprecated
    INSERTING: 15,             // storage serv only
    INVALID_NAME_SPACE: 16,    // directory serv only
    ALREADY_FULFILLED: 17,     // storage serv only
    NOT_SAME_RECORD: 18,       // storage serv only
    VERSION_CONFLICT: 19       // storage serv replicated data inconsistency
};

module.exports = {
    MessageSenderType: MessageSenderType,
    CONSTANTS: CONSTANTS,
    Direction: Direction,
    DataTags: DataTags,
    OperationType: OperationType,
    OperationStatus: OperationStatus
};
