// consumerweb-logger/lib/bizevents-logger.js
'use strict';

/**
 * Utility library to send biz events to CAL.
 * These are custom biz events for getting business metrics.
 * More details here : https://confluence.paypal.com/cnfl/display/CAL/CAL+BIZ+Guidelines
 * Usage :
 * For example to send BIZ events for successful add card.
 * logger.logBizEvent(req, logger.bizEventEnums.ADD_CARD, logger.bizEventEnums.SUCCESS);
 *
 * @author: suranjan@paypal.com
 */

var cal = require('cal');

//Constants used in biz event logging
var EVENT_TYPE = 'BIZ',
	BIZ_SUCCESS = cal.Status.SUCCESS,
	BIZ_ERROR = cal.Status.ERROR;

/**
 * Log a business event to cal.
 * @param req request object
 * @param name of cal biz event eg. addCardMetrics
 * @param status of cal biz event
 * @param data is custom data
 */
function logBizEvent(req, name, status, customData) {
	var calEvent,
		locality = req && req.locality,
		locale = locality && locality.locale,
		country = locality && locality.country;

	if (cal) {
		calEvent = cal.createEvent(EVENT_TYPE, name, status);
		calEvent.status = status || cal.Status.SUCCESS;

		customData = customData || {};

		// set correlation id.
		customData.corr_id = req.correlationId;
		// set country.
		customData.cntry = country;
		// set locale.
		customData.locale = locale;

		// set custom data.
		calEvent.addData(customData);

		// send cal biz event.
		calEvent.complete();
	}
}

var bizEventEnums = {
	AUTH_FLOW: 'authFlowMetrics',
	ADD_CARD: 'addCardMetrics',
	ADD_BANK: 'addBankMetrics',
	CONFIRM_CARD: 'confirmCardMetrics',
	CONFIRM_BANK: 'confirmBankMetrics',
	DELETE_CARD: 'deleteCardMetrics',
	DELETE_BANK: 'deleteBankMetrics',
	AUTHORIZE_BANK: 'authorizeBankMetrics',
	SET_BANK_PREFERENCE: 'setBankPreferenceMetrics',
	UPDATE_CARD: 'updateCardMetrics',
	ANALYZE_WITHDRAW_FUNDS: 'analyzeWithdrawFundsMetrics',
	ANALYZE_ADD_FUNDS: 'analyzeAddFundsMetrics',
	WITHDRAW_FUNDS: 'withdrawFundsMetrics',
	ADD_FUNDS: 'addFundsMetrics',
	TRANSFER: {
		GET_FI: 'transferGetFIMetrics',
		REQUEST_MONEY: 'transferRequestMoneyMetrics',
		SEND_MONEY: 'transferSendMoneyMetrics'
	},
	GET_CARD_BALANCE: 'getCardBalance',
	SUCCESS: BIZ_SUCCESS,
	ERROR: BIZ_ERROR
};

module.exports = {
	logBizEvent: logBizEvent,
	bizEventEnums: bizEventEnums
};
