// servicecore/lib/transport/generic.js
'use strict';

var httpunch = require('httpunch'),
    utils = require('../utils');

/**
 * Generic transport is just a passthrough to httpunch (regular http request) with JSON parsing for JSON content-type.
 * @type {exports}
 */
module.exports = function () {
    return function transport(options, callback) {
        var req;

        req = utils.nsBindRequest(httpunch.request({
            protocol: options.protocol,
            method: options.method,
            headers: options.headers,
            hostname: options.hostname,
            port: options.port,
            path: options.path,
            qs: options.qs,
            pathparams: options.pathparams,
            body: options.body,
            socketTimeout: options.socketTimeout,
            connectTimeout: options.connectTimeout,
            maxRedirects: options.maxRedirects,
            agent: options.agent
        }));

        req.end(function handleResponse(error, response) {
            if (error) {
                callback(error);
                return;
            }

            if (utils.isJSON(response)) {
                response.body = utils.tryParse(response.body);
            }

            callback.apply(null, arguments);
        });

        return req;
    };
};
