// nodejs-ci-suite/lib/template/gruntfile.js
'use strict';

var path = require('path'),
	nconf = require('nconf'),
	confer = require('confer');

//Nconf Read utility function
function read(prop, isArray) {
	if (isArray) {
		return nconf.get(prop).split(',');
	}
	return  nconf.get(prop);
}

module.exports = function (grunt) {
	//Strip the color option from logging. Suitable for logging to build output.
	grunt.option('color', false);
	
	//Load properties in the order of (1) Command line  (2) Environment Variables , and (3) Json properties file
	nconf.argv()
		.env()
		.file({ file: 'ci-build.json'});
	
	
	var jsHintrcFile	= read('JSHINT_RC'),//The JsHintrc file location
		jsHintJSON		= confer(path.resolve(jsHintrcFile)) || {},
		jsHintReporter	= read('JSHINT_REPORTER'),//The JsHint Reporter
		jshintSrc		= read('JSHINT_SRC', true),//The files to run Linting
		checkstyle		= read('CHECKSTYLE_FILE'),//The Checkstyle.xml path 
		reportDir		= read('REPORTS_DIR'),//The Reports Home directory
		testFiles		= read('TEST_FILES', true),//The files to run mocha test
		platoReport		= read('REPORTS_PLATO'),//The Plato Report file location
		covReport		= read('REPORTS_COVERAGE'),//The Coverage Report file location 
		covType			= read('COVERAGE_TYPE'),//The Coverage Report type
		covPrintType	= read('COVERAGE_PRINT_TYPE'),//The Coverage print type
		covJsonFile		= read('COVERAGE_FILE'),//The Coverage.json file location
		covSrc			= read('COVERAGE_SRC', true), //The Files to run coverage test
		mochaReporter	= read('MOCHA_REPORTER'), //The Reporter type
		cwd				= process.cwd();
			
	jsHintJSON.globals = {
		exports : true
	};
	//The CheckStyle Environment Variable
	if (checkstyle) {
		process.env.CHECKSTYLE_FILE = path.resolve(checkstyle);
	}
	// Project configuration.
	grunt.initConfig({
		pkg : '<json:package.json>',
		chdir :  {
			application : {
				src	: cwd
			}
		},
		clean :  {
			reports : {
				src	: reportDir
			}
		},
		//JS Hint configuration
		jshint : {
			all : {
				src : jshintSrc,
				options : {
					jshintrc : jsHintrcFile,
					reporter : jsHintReporter
				},
				globals : {
					exports : true
				}
			}
		},
		//Mocha configuration
		mochatest : {
			all : {
				src: testFiles,
				options: {
					globals: ['should'],
					timeout: 15000,
					ignoreLeaks: false,
					ui: 'bdd',
					reporter: mochaReporter
				}
			}
		},
		//Plato configuration
		plato : {
			all : {
				src : jshintSrc,
				dest : platoReport,
				options : {
					jshint : jsHintJSON
				}
			}
		},
		//Istanbul configurations
		codecoverage : {
			all : {
				src: testFiles,
				options: {
					globals: ['should'],
					timeout: 3000,
					ignoreLeaks: false,
					ui: 'bdd',
					reporter: mochaReporter,
					covDir: covReport,			//Istanbul option
					reportType: covType,		//Istanbul option
					printType: covPrintType		//Istanbul option
				}
			}
		}
		
	});
	
	//Load tasks
	grunt.loadNpmTasks('grunt-contrib-clean');
	grunt.loadNpmTasks('grunt-plato');
	//grunt.loadNpmTasks('grunt-istanbul');
	//CI suite
	grunt.loadNpmTasks('grunt-ci-suite');
	
	//Test - JsHint and Mocha
	grunt.registerTask('test', ['jshint', 'mochatest']);
	//Coverage - Plato and Istanbul coverage
	grunt.registerTask('coverage', [
		'chdir',
		'clean',
		'plato',
		'codecoverage'
	]);
	// Default task.
	grunt.registerTask('default', ['test']);
	
};
