// vobin/lib/constants.js
'use strict';

var defaultVersion = module.exports.defaultVersion = 0x01;

// Universal Constants
var UNIVERSAL_HEADER_BYTE = module.exports.UNIVERSAL_HEADER_BYTE = 0x55; // 'U' Used to mark this as a universal format.
var ASCII_ZERO = module.exports.ASCII_ZERO = 0x30; // The ascii character zero '0'
var FLAG_USE_BASE64 = module.exports.FLAG_USE_BASE64 = 0x01; // Flag we use if base64 is on.
var FLAG_USE_COMPRESSION = module.exports.FLAG_USE_COMPRESSION = 0x02; // Flag we use if compression is on.

// Universal Format Ids
var BINARY_FORMAT_ID = module.exports.BINARY_FORMAT_ID = 2 + ASCII_ZERO;
var COMPRESSEDBINARY_FORMAT_ID = module.exports.COMPRESSEDBINARY_FORMAT_ID = 3 + ASCII_ZERO;

// Header Constants
var COMPRESSEDBINARY_HEADER_1 = module.exports.COMPRESSEDBINARY_HEADER_1 = 0x8B; // The header byte denoting compressed binary
var COMPRESSEDBINARY_HEADER_2 = module.exports.COMPRESSEDBINARY_HEADER_2 = 0x6B; //

var BINARY_HEADER = module.exports.BINARY_HEADER = new Buffer(2); // "VO"
BINARY_HEADER[0] = 0x56;
BINARY_HEADER[1] = 0x4F;
