// vobin/lib/types.js
'use strict';

var NEW_OBJ = module.exports.NEW_OBJ = 0x01;
var END_OBJ = module.exports.END_OBJ = 0x02;
var OLD_OBJ = module.exports.OLD_OBJ = 0x06;
var NEW_FIELD = module.exports.NEW_FIELD = 0x03;
var OLD_FIELD = module.exports.OLD_FIELD = 0x04;
var END_FIELD = module.exports.END_FIELD = 0x05;
var OBJ_POINTER = module.exports.OBJ_POINTER = 0x07;

var PRIMITIVE = module.exports.PRIMITIVE = 0x01;
var OBJ_FIELD = module.exports.OBJ_FIELD = 0x02;
var ARRAY = module.exports.ARRAY = 0x03;
var HASHMAP = module.exports.HASHMAP = 0x04;
var POINTER = module.exports.POINTER = 0x05;

var BOOLEAN = module.exports.BOOLEAN = 0x01;
var CHAR = module.exports.CHAR = 0x02;
var SINT8 = module.exports.SINT8 = 0x02;
var UINT8 = module.exports.UINT8 = 0x03;
var SHORT = module.exports.SHORT = 0x04;
var SINT16 = module.exports.SINT16 = 0x04;
var UINT16 = module.exports.UINT16 = 0x05;
var USHORT = module.exports.USHORT = 0x05;
var SINT32 = module.exports.SINT32 = 0x06;
var INT = module.exports.INT = 0x06;
var LONG = module.exports.LONG = 0x06;
var UINT32 = module.exports.UINT32 = 0x07;
var UINT = module.exports.UINT = 0x07;
var ULONG = module.exports.ULONG = 0x07;
var SINT64 = module.exports.SINT64 = 0x08;
var LLONG = module.exports.LLONG = 0x08;
var UINT64 = module.exports.UINT64 = 0x09;
var ULLONG = module.exports.ULLONG = 0x09;
var FLOAT = module.exports.FLOAT = 0x0a;
var DOUBLE = module.exports.DOUBLE = 0x0b;
var STRING = module.exports.STRING = 0x0c;
var CURRENCY = module.exports.CURRENCY = 0x0d;
var BUFFER = module.exports.BUFFER = 0x0e;

var NEG_SIGN = module.exports.NEG_SIGN = 0xe1;
var NULL_BYTE = module.exports.NULL_BYTE = 0x00;

var typeMap = module.exports.typeMap = {};

typeMap.bool = BOOLEAN;
typeMap.char = CHAR;
typeMap.int = INT;
typeMap.uint = UINT;
typeMap.short = SHORT;
typeMap.ushort = USHORT;
typeMap.long = LONG;
typeMap.ulong = ULONG;
typeMap.llong = LLONG;
typeMap.ullong = ULLONG;
typeMap.float = FLOAT;
typeMap.double = DOUBLE;
typeMap.String = STRING;
typeMap.Currency = CURRENCY;
typeMap.Buffer = BUFFER;
typeMap.string = STRING;
typeMap.currency = CURRENCY;
typeMap.buffer = BUFFER;
typeMap.sint8 = SINT8;
typeMap.uint8 = UINT8;
typeMap.sint16 = SINT16;
typeMap.uint16 = UINT16;
typeMap.sint32 = SINT32;
typeMap.uint32 = UINT32;
typeMap.sint64 = SINT64;
typeMap.uint64 = UINT64;

var typeReverseMap = module.exports.typeReverseMap = {};

typeReverseMap[BOOLEAN] = 'bool';
typeReverseMap[CHAR] = 'char';
typeReverseMap[INT] = 'int';
typeReverseMap[UINT] = 'uint';
typeReverseMap[SHORT] = 'short';
typeReverseMap[USHORT] = 'ushort';
typeReverseMap[LONG] = 'long';
typeReverseMap[ULONG] = 'ulong';
typeReverseMap[LLONG] = 'llong';
typeReverseMap[ULLONG] = 'ullong';
typeReverseMap[FLOAT] = 'float';
typeReverseMap[DOUBLE] = 'double';
typeReverseMap[STRING] = 'String';
typeReverseMap[CURRENCY] = 'Currency';
typeReverseMap[BUFFER] = 'Buffer';
typeReverseMap[STRING] = 'string';
typeReverseMap[CURRENCY] = 'currency';
typeReverseMap[BUFFER] = 'buffer';
typeReverseMap[SINT8] = 'sint8';
typeReverseMap[UINT8] = 'uint8';
typeReverseMap[SINT16] = 'sint16';
typeReverseMap[UINT16] = 'uint16';
typeReverseMap[SINT32] = 'sint32';
typeReverseMap[UINT32] = 'uint32';
typeReverseMap[SINT64] = 'sint64';
typeReverseMap[UINT64] = 'uint64';

var typeJsonMap = module.exports.typeJsonMap = [];

typeJsonMap[BOOLEAN] = 'boolean';
typeJsonMap[CHAR] = 'string';
typeJsonMap[INT] = 'number';
typeJsonMap[UINT] = 'number';
typeJsonMap[SHORT] = 'number';
typeJsonMap[USHORT] = 'number';
typeJsonMap[LONG] = 'number';
typeJsonMap[ULONG] = 'number';
typeJsonMap[FLOAT] = 'number';
typeJsonMap[DOUBLE] = 'number';
typeJsonMap[STRING] = 'string';
typeJsonMap[CURRENCY] = 'string';
typeJsonMap[BUFFER] = 'object';
typeJsonMap[STRING] = 'string';
typeJsonMap[SINT8] = 'number';
typeJsonMap[UINT8] = 'number';
typeJsonMap[SINT16] = 'number';
typeJsonMap[UINT16] = 'number';
typeJsonMap[SINT32] = 'number';
typeJsonMap[UINT32] = 'number';

typeJsonMap[SINT64] = typeJsonMap[UINT64] = typeJsonMap[LLONG] = typeJsonMap[ULLONG] = function (value) {
    if (typeof value !== 'string') {
        if (value >= -0x1fffffff && value <= 0x1fffffff) {
            return 'string|number';
        }
    }
    return 'string';
};

module.exports.isVO = function (type) {
    return !typeMap[type];
};
