// objutil/index.js
'use strict';

var util = require('util'),
    thing = require('core-util-is');

/**
 * Copies properties from src to dest, potentially overwriting previously defined properties.
 * @param src
 * @param dest
 * @returns {*}
 */
exports.mixin = function mixin(src, dest) {
    var sources = Array.prototype.slice.call(arguments);
    dest = sources.pop();

    sources.forEach(function (source) {
        if (thing.isObject(source)) {
            Object.getOwnPropertyNames(source).forEach(function(name) {
                Object.defineProperty(dest, name, Object.getOwnPropertyDescriptor(source, name));
            });
        }
    });
    return dest;
};


/**
 * Copies values to dest *ONLY* if they were not previously defined.
 * @param src
 * @param dest
 */
exports.merge = function merge(src, dest) {
    var sources = Array.prototype.slice.call(arguments);
    dest = sources.pop();

    sources.forEach(function (source) {
        if (thing.isObject(source)) {
            Object.getOwnPropertyNames(source).forEach(function(name) {
                if (dest[name] === undefined) {
                    Object.defineProperty(dest, name, Object.getOwnPropertyDescriptor(source, name));
                }
            });
        }
    });
    return dest;
};


exports.extend = function extend(ctor, superCtor, proto) {
    util.inherits(ctor, superCtor);
    if (proto && typeof proto === 'object') {
        exports.mixin(proto, ctor.prototype);
    }
};


exports.clone = exports.copy = function clone(src) {
    var dest = src;

    if (thing.isObject(src)) {
        dest = Array.isArray(src) ? [] : Object.create(Object.getPrototypeOf(src));
        Object.getOwnPropertyNames(src).forEach(function (prop) {
            Object.defineProperty(dest, prop, Object.getOwnPropertyDescriptor(src, prop));
        });
    }

    return dest;
};


exports.deepClone = exports.deepCopy = function deepClone(src) {
    var dest = src;

    if (thing.isObject(src)) {
        dest = Array.isArray(src) ? [] : Object.create(Object.getPrototypeOf(src));
        Object.getOwnPropertyNames(src).forEach(function (prop) {
            var descriptor = Object.getOwnPropertyDescriptor(src, prop);
            descriptor.value = deepClone(descriptor.value);
            Object.defineProperty(dest, prop, descriptor);
        });
    }

    return dest;
};
