// topos/lib/bibliography.js
'use strict';

var path = require('path');

var serviceTokenPrefix = 'https://86c873e2ac25756b58f015f4390641a6f554ccfe:x-oauth-basic';
var repoPrefix = '@github.paypal.com/raw/Topology/topo/next-next-release_int';

var dir = exports.dir = {
    cdbs:      path.join(process.cwd(), 'cdbs'),
    topos:     path.join(__dirname, '..', 'topos'),
    template:  path.join(__dirname, '..', 'template')
};

exports.file = {

    appdef:    path.join(dir.cdbs, 'application.def'),
    apptext:   path.join(dir.cdbs, 'application.txt'),
    applayout: path.join(dir.cdbs, 'application.layout'),
    layout:    path.join(dir.cdbs, 'layout.json'),
    layoutrel: path.join('cdbs', 'layout.json'),//Use relative path
    layouttemplate:    path.join(dir.cdbs, 'layout_template.json'),
    layouttemplaterel: path.join('cdbs', 'layout_template.json'),//Use relative path
    manifest:  path.join(process.cwd(), 'service-manifest.yaml'),

    topo: {
        main:    path.join(dir.topos, 'topo.main'),
        stageEnv: path.join(dir.topos, 'stage.env'),
        stage2Env: path.join(dir.topos, 'stage2.env'),
        layout:  path.join(dir.topos, 'layout.json'),
        ports:   path.join(dir.topos, 'ports.json')
    },

    template: {
        appdef: path.join(dir.template, 'application.def')
    }
};

exports.url = {
    topo: {
        main: serviceTokenPrefix + repoPrefix + '/topo/topo.main',
        stageEnv: serviceTokenPrefix + repoPrefix + '/topo/stage.env',
        stage2Env: serviceTokenPrefix + repoPrefix + '/topo/stage2.env'
    }
};
